lib/staging/tls/tls.cpp

```diff
@@ -692,6 +697,36 @@ std::uint32_t ServerConnection::m_count{0};
 std::mutex ServerConnection::m_cv_mutex;
 std::condition_variable ServerConnection::m_cv;
 
+namespace {
+
+std::unique_ptr<std::filesystem::path> keylog_file;
+std::unique_ptr<TlsKeyLoggingServer> keylog_server;
+
```

Comment:
I think SSL session variables should be used rather than file scoped variables.
There is a chance that keylog_server/keylog_file change before or during the execution of keylog_callback().

CRYPTO_get_ex_new_index(CRYPTO_EX_INDEX_SSL, 0, nullptr, nullptr, nullptr, nullptr)
SSL_set_ex_data(ctx, s_index, ptr)
SSL_get_ex_data

These are used in extensions/trusted_ca_keys.cpp

Reply:
Looking at the code the log file is the same for all connections and can be set in the common SSL context.
The log server seems to be created for every connection and could be created as part of the ServerConnection object and a pointer made available to the callback as a SSL session variable. Thus the keylog_server would be consistent for the whole lifetime of the connection.
This would support sending out the new information when a TLS connection updates its keys.

Reply:
Switched to `SSL_set_ex_data`, `SSL_get_ex_data`